clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -402,26 +403,38 @@ public void refreshCommittedOffsetsIfNeeded() {
         }
     }
 
-    @Override
-    public void close() {
+    public void close(long timeoutMs, long requestTimeoutMs) {
         // we do not need to re-enable wakeups since we are closing already
         client.disableWakeups();
+        long startTimeMs = time.milliseconds();
         try {
-            maybeAutoCommitOffsetsSync();
+            timeoutMs = Math.min(timeoutMs, requestTimeoutMs);
+            maybeAutoCommitOffsetsSync(timeoutMs);
+        } finally {
+            super.close();
```

Comment:
@hachikuji @apurvam Thank you for the reviews. I think I understand - I have redone the previous change. Can you take a look at the latest commit, please? I think `commitSync` doesn't need any special treatment for wakeup because if the consumer was interrupted in a poll while committing, the final loop in close will take care of sending the commit and awaiting response as long as the coordinator is known. Async commit needs special handling since the commit might not have been queued due a rebalance. So that needs to find coordinator first. I have made that code separate from the final loop. `pendingAsyncCommits` only tracks async commits that are waiting to find coordinator to queue the commit request. Once the requests are queued (in the first part of close, commits followed by LeaveGroup are queued), the loop just does the processing of queued requests as long as coordinator is not null.